plugin/trino-hudi/src/main/java/io/trino/plugin/hudi/query/index/HudiPartitionStatsIndexSupport.java

```diff
@@ -0,0 +1,102 @@
+
+import io.airlift.log.Logger;
+import io.trino.plugin.hudi.util.TupleDomainUtils;
+import io.trino.spi.predicate.TupleDomain;
+import org.apache.hudi.avro.model.HoodieMetadataColumnStats;
+import org.apache.hudi.common.model.FileSlice;
+import org.apache.hudi.common.table.HoodieTableMetaClient;
+import org.apache.hudi.common.util.hash.ColumnIndexID;
+import org.apache.hudi.metadata.HoodieTableMetadata;
+import org.apache.hudi.metadata.HoodieTableMetadataUtil;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import static io.trino.plugin.hudi.util.TupleDomainUtils.hasSimpleNullCheck;
+
+public class HudiPartitionStatsIndexSupport
+        extends HudiColumnStatsIndexSupport
+{
+    private static final Logger log = Logger.get(HudiColumnStatsIndexSupport.class);
+
+    public HudiPartitionStatsIndexSupport(HoodieTableMetaClient metaClient)
+    {
+        super(log, metaClient);
+    }
+
+    @Override
+    public Map<String, List<FileSlice>> lookupCandidateFilesInMetadataTable(HoodieTableMetadata metadataTable, Map<String, List<FileSlice>> inputFileSlices, TupleDomain<String> regularColumnPredicates)
+    {
+        return Map.of();
+    }
+
+    public Optional<List<String>> prunePartitions(HoodieTableMetadata metadataTable, TupleDomain<String> regularColumnPredicates)
+    {
+        // Filter out predicates containing simple null checks (`IS NULL` or `IS NOT NULL`)
+        TupleDomain<String> filteredRegularPredicates = regularColumnPredicates.filter((_, domain) -> !hasSimpleNullCheck(domain));
+
+        // Sanity check, if no regular domains, return immediately
+        if (filteredRegularPredicates.getDomains().isEmpty()) {
+            return Optional.empty();
+        }
+
+        List<String> regularColumns = new ArrayList<>(filteredRegularPredicates.getDomains().get().keySet());
+
+        // Get columns to filter on
+        List<String> encodedTargetColumnNames = regularColumns
+                .stream()
+                .map(col -> new ColumnIndexID(col).asBase64EncodedString()).toList();
+
+        // Map of partition stats keyed by partition name
+        // If some columns in encodedTargetColumnNames is not available in partition stats index, partition will not be pruned
+        // This is so as we cannot be sure that the constraint is true or not, hence, adopt conservative measure
+        Map<String, List<HoodieMetadataColumnStats>> statsByPartitionName = metadataTable.getRecordsByKeyPrefixes(
+                        encodedTargetColumnNames,
+                        HoodieTableMetadataUtil.PARTITION_NAME_PARTITION_STATS, true)
+                .collectAsList()
+                .stream()
+                .filter(f -> f.getData().getColumnStatMetadata().isPresent())
+                .map(f -> f.getData().getColumnStatMetadata().get())
+                .collect(Collectors.groupingBy(HoodieMetadataColumnStats::getFileName));
```

Comment:
If the key of map is partition name, then grouping should be done based on that instead of file name right?

Reply:
I was also confused by this initially. The filename is the partition name for partition stats index.